google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/ReadOnlyTransactionTest.java

```diff
@@ -0,0 +1,403 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner.jdbc;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.CoreMatchers.nullValue;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.cloud.Timestamp;
+import com.google.cloud.spanner.DatabaseClient;
+import com.google.cloud.spanner.ErrorCode;
+import com.google.cloud.spanner.Key;
+import com.google.cloud.spanner.KeySet;
+import com.google.cloud.spanner.Mutation;
+import com.google.cloud.spanner.Options;
+import com.google.cloud.spanner.Options.QueryOption;
+import com.google.cloud.spanner.Options.ReadOption;
+import com.google.cloud.spanner.ResultSet;
+import com.google.cloud.spanner.SpannerException;
+import com.google.cloud.spanner.Statement;
+import com.google.cloud.spanner.Struct;
+import com.google.cloud.spanner.TimestampBound;
+import com.google.cloud.spanner.jdbc.StatementParser.ParsedStatement;
+import com.google.cloud.spanner.jdbc.StatementParser.StatementType;
+import com.google.cloud.spanner.jdbc.UnitOfWork.UnitOfWorkState;
+import com.google.spanner.v1.ResultSetStats;
+import java.util.Arrays;
+import java.util.Calendar;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class ReadOnlyTransactionTest {
+  @Rule public ExpectedException exception = ExpectedException.none();
+
+  private static final class SimpleReadOnlyTransaction
+      implements com.google.cloud.spanner.ReadOnlyTransaction {
+    private Timestamp readTimestamp = null;
+    private final TimestampBound staleness;
+
+    private SimpleReadOnlyTransaction(TimestampBound staleness) {
+      this.staleness = staleness;
+    }
+
+    @Override
+    public ResultSet read(
+        String table, KeySet keys, Iterable<String> columns, ReadOption... options) {
+      return null;
+    }
+
+    @Override
+    public ResultSet readUsingIndex(
+        String table, String index, KeySet keys, Iterable<String> columns, ReadOption... options) {
+      return null;
+    }
+
+    @Override
+    public Struct readRow(String table, Key key, Iterable<String> columns) {
+      return null;
+    }
+
+    @Override
+    public Struct readRowUsingIndex(String table, String index, Key key, Iterable<String> columns) {
+      return null;
+    }
+
+    @Override
+    public ResultSet executeQuery(Statement statement, QueryOption... options) {
+      if (readTimestamp == null) {
```

Comment:
How come we are overriding these methods for the test?

Reply:
I guess the naming is a little bit confusing in this case, as we have two different `ReadOnlyTransaction`s in this case. The interface that is being implemented is `com.google.cloud.spanner.ReadOnlyTransaction`, i.e. the read-only transaction of the normal client library, and not the read-only transaction of the connection API. That is done in this test in order to create a simple mock object with some logic to use in the test cases.

The logic in the `executeQuery(...)` method is intended to make sure the connection API actually uses the correct staleness mode for the underlying read-only transaction (and to return a mock result set).